internal/webhook/external/bitbucket.go

```diff
@@ -0,0 +1,212 @@
+func (b *bitbucketWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[bitbucketSecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for a Bitbucket WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// GetHandler implements WebhookReceiver.
+func (b *bitbucketWebhookReceiver) GetHandler() http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		logger := logging.LoggerFromContext(ctx)
+
+		signingKey, ok := b.secretData[bitbucketSecretDataKey]
+		if !ok {
+			xhttp.WriteErrorJSON(w, nil)
+			return
+		}
+
+		eventType := r.Header.Get("X-Event-Key")
+		switch eventType {
+		case bitbucketPushEvent:
+		default:
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					fmt.Errorf("event type %s is not supported", eventType),
+					http.StatusNotImplemented,
+				),
+			)
+			return
+		}
+
+		// Early check of Content-Length if available
+		if contentLength := r.ContentLength; contentLength > bitbucketWebhookBodyMaxBytes {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					fmt.Errorf("content exceeds limit of %d bytes", bitbucketWebhookBodyMaxBytes),
+					http.StatusRequestEntityTooLarge,
+				),
+			)
+			return
+		}
+
+		body, err := io.LimitRead(r.Body, bitbucketWebhookBodyMaxBytes)
+		if err != nil {
+			if errors.Is(err, &io.BodyTooLargeError{}) {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(err, http.StatusRequestEntityTooLarge),
+				)
+				return
+			}
+			xhttp.WriteErrorJSON(w, err)
+			return
+		}
+
+		sig := r.Header.Get("X-Hub-Signature")
+		if sig == "" {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(errors.New("missing signature"), http.StatusUnauthorized),
+			)
+			return
+		}
+
+		// Note: github.com/google/go-github/v71/github has a great implementation
+		// of HMAC signature validation that isn't GitHub-specific, so we've opted
+		// to use it here for Bitbucket as well.
+		if err = gh.ValidateSignature(sig, body, signingKey); err != nil {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(errors.New("unauthorized"), http.StatusUnauthorized),
+			)
+			return
+		}
+
+		payload := struct {
+			Repository struct {
+				Links struct {
+					HTML struct {
+						Href string `json:"href"`
+					} `json:"html"`
+				} `json:"links"`
+			} `json:"repository"`
+		}{}
+		if err = json.Unmarshal(body, &payload); err != nil {
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(errors.New("invalid request body"), http.StatusBadRequest),
+			)
+			return
+		}
+
+		// Note: It may seem peculiar to obtain the repo URL from the payload's
+		// repository.links.html.href field, however, there is no better option.
+		//
+		// A naive option would be combining https://bitbucket.org/ with the value
+		// of the payload's repository.full_name field, but that does not hold up
+		// for events originating from BitBucket Data Center, which will utilize a
```

Comment:
This was my bad. They don't capitalize the second b.

```suggestion
		// for events originating from Bitbucket Data Center, which will utilize a
```